### PR TITLE
Resource: Unmanaged Solution

### LINKED ISSUE #1117: Add unmanaged solution resource
## User Story

As a **Terraform user**, I want to manage **unmanaged Dataverse solutions** through Terraform so that I can **declare stable solution identity and metadata in IaC without coupling lifecycle to a ZIP deployment artifact**.

Use case: platform teams need to create and manage empty unmanaged solutions as first-class objects for ALM, Git integration, and environment bootstrapping. This is valuable when creating solutions up front for source control binding, environment scaffolding, and repeatable platform provisioning workflows.

## Resource

- **Resource Name:** `powerplatform_unmanaged_solution`
- **Service Name:** `solution`
- **Documentation Link:**
  - [solution table reference](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/reference/solution?view=dataverse-latest)

### Potential Terraform Configuration

```hcl
resource "powerplatform_unmanaged_solution" "example" {
  environment_id = powerplatform_environment.example.id
  uniquename     = "MySolution"
  display_name   = "My Solution"
  publisher_id   = "f9379a7f-a5a8-f011-bbd3-000000000000"
  description    = "Solution description text example here."
}
```

### Additional Validation Rules

- `environment_id` must reference an environment with Dataverse enabled.
- `uniquename` must be unique within the target environment.
- `publisher_id` must reference an existing publisher in the target Dataverse environment.
- `display_name` should be required and treated as user-managed metadata.
- `description` should be optional.
- `solution_version` should **not** be exposed as a user-managed lifecycle property for this resource.

## API documentation

| Action | Verb | URL | Status Codes | Comments |
|--------|------|-----|--------------|----------|
| Create | POST | `/api/data/v9.0/solutions` | 201 | Create unmanaged solution directly without ZIP import |
| Read   | GET  | `/api/data/v9.2/solutions?$expand=publisherid&$filter=solutionid eq {id}` | 200 | |
| Update | PATCH | `/api/data/v9.0/solutions({id})` | 204 | Update mutable metadata such as `friendlyname` / `description` if supported |
| Delete | DELETE | `/api/data/v9.2/solutions({id})` | 204 | Delete unmanaged solution |

### JSON

```json
{
  "description": "Solution description text example here.",
  "friendlyname": "My Solution",
  "publisherid@odata.bind": "/publishers(f9379a7f-a5a8-f011-bbd3-000000000000)",
  "uniquename": "MySolution",
  "version": "1.0.0.0"
}
```

## Why a new resource is needed

The current `powerplatform_solution` resource is centered on importing a `solution.zip`, rather than modeling the underlying Dataverse solution object. This causes it to behave more like a deployment/import operation than a Terraform resource representing the Dataverse solution object itself.

This creates several Terraform modeling problems:

- guaranteed drift for unmanaged solutions, because the lifecycle is tied to ZIP import rather than stable object reconciliation
- non-idempotent behavior, because repeated applies are shaped around import/package actions rather than stable object reconciliation
- lifecycle mismatch, because ZIP deployment is an operational action while the Terraform resource should model the long-lived solution object

This separation would allow Terraform to model Dataverse solutions in a way that is both idiomatic and aligned with the platform's actual object model and lifecycle behavior.

## Existing resource naming / scope

The existing `powerplatform_solution` resource appears to be much more appropriate for managed solution package import semantics than for unmanaged solution lifecycle semantics.

Because of that, it would be clearer if:

- a new `powerplatform_unmanaged_solution` resource is added for direct solution object management
- the current ZIP-based `powerplatform_solution` resource is renamed to `powerplatform_managed_solution`

That would better reflect the actual semantics:

- `powerplatform_unmanaged_solution`
  - manages the Dataverse solution object itself
- `powerplatform_managed_solution`
  - manages ZIP/package import lifecycle

## Alternative design

An alternative would be to keep a single unified `powerplatform_solution` resource and redesign it so that:

- the solution object identity is always first-class
- ZIP/package import becomes an optional nested block (or even another related resource), for example `managed_package`

Example direction:

```hcl
resource "powerplatform_solution" "example" {
  environment_id = powerplatform_environment.example.id
  uniquename     = "mysolution"
  display_name   = "My Solution"
  publisher_id   = "f9379a7f-a5a8-f011-bbd3-000000000000"

  managed_package {
    solution_file = "solution_managed.zip"
    settings_file = "settings.json"
  }
}
```

This would keep the solution abstraction unified, but would likely be a breaking change from the current resource contract because the existing `powerplatform_solution` resource is centered on ZIP import rather than explicit solution identity.

One reason this alternative is still plausible is that a `managed_package` block behaves more like an acceptable target-state declaration than unmanaged ZIP import does. In practice, Terraform and Power Platform can often treat managed package installation as a reconciliation attempt against the environment, rather than as identity definition.

## Solution Import/Install Constraints

A fundamental constraint we need to acknowledge is the real Dataverse semantics around solution package deployment:

- non-deterministic imports
- non-deterministic idempotent re-apply behavior
- no safe deterministic downgrade support
- hidden state and history inside Dataverse/import jobs

So regardless of which packaged zip option is more acceptable, it is still fundamentally deployment-oriented behavior and should be modeled with care. These constraints also mean it is not a fully safe Terraform resource candidate in the same way that a normal CRUD-style infrastructure object would be.

## Version handling

Although the create API requires an initial version in the request body, that does not mean version should be modeled as a user-managed Terraform property for unmanaged solution identity.

A provider implementation can simply send:

- `version = "1.0.0.0"` during create

without exposing solution version as part of the unmanaged solution resource contract.

## Definition of Done

- [x] Data Transfer Objects (dtos) in `dto.go`
- [x] Resource Model in `model.go`
- [x] API Client functions in `api_solution.go`
- [x] Resource Implementation in `resource_unmanaged_solution.go`
- [x] Unit Tests in `resource_unmanaged_solution_test.go` for Happy Path, Error conditions, boundary cases
- [x] Acceptance Tests in `resource_unmanaged_solution_test.go` for Happy Path
- [x] Resource Added to `provider.go` and `provider_test.go`
- [x] Example in the `/examples` folder
- [x] Schema documented using `MarkdownDescription`
- [x] Change log entry `changie new -k added`
- [x] Run `make precommit` before PR

See the [contributing guide](https://github.com/microsoft/terraform-provider-power-platform/blob/main/CONTRIBUTING.md) for more information about what's expected for contributions.

### PR DESCRIPTION
**Pull request by AdamCoulterOz**
## Summary
- add `powerplatform_unmanaged_solution` as a first-class Dataverse CRUD resource
- add `powerplatform_unmanaged_solution` data source for lookup by environment and unique name
- add unit and acceptance coverage plus generated docs and examples

## Testing
- go test ./internal/services/solution ./internal/provider -run 'TestUnit(UnmanagedSolutionResource_Validate_Create|UnmanagedSolutionResource_Validate_Update|UnmanagedSolutionResource_Validate_Create_No_Dataverse|UnmanagedSolutionResource_Validate_Create_Eventual_Consistency|UnmanagedSolutionResource_Validate_Create_Uses_Response_ID|UnmanagedSolutionDataSource_Validate_Read|UnmanagedSolutionDataSource_Validate_No_Dataverse|UnmanagedSolutionDataSource_Validate_Not_Found|PowerPlatformProviderHasChildDataSources_Basic|PowerPlatformProviderHasChildResources_Basic)$'
- PATH=/Users/adam/go/bin:$PATH make userdocs
- POWER_PLATFORM_USE_CLI=true TF_ACC=1 go test ./internal/services/solution -run ^TestAccUnmanagedSolutionResource_Validate_Create_Update -timeout 90m

Closes #1117

### CODE REVIEW

File: internal/services/solution/datasource_unmanaged_solution.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
```
Comment (mawasile) [inline internal/services/solution/datasource_unmanaged_solution.go]: do we need datasource since we already have `powerplatform_solutions` in the provider doing the same thing?

File: internal/services/solution/resource_unmanaged_solution.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+func validateUnmanagedSolution(solution *SolutionDto, typeName string) error {
+	if solution != nil && solution.IsManaged {
+		return fmt.Errorf("solution '%s' is managed and cannot be used with %s", solution.Name, typeName)
+	}
+
+	return nil
+}
+
+func splitSolutionCompositeID(id string) []string {
+	return strings.SplitN(id, "_", 2)
```
Comment (mawasile) [inline internal/services/solution/resource_unmanaged_solution.go]:
```suggestion
	return strings.SplitN(id, "/", 2)
```

**Review by mawasile — CHANGES_REQUESTED** [review]

File: internal/services/solution/resource_unmanaged_solution_mapper_test.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package solution
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/resource_unmanaged_solution_mapper_test.go]: Addressed - that mapper test file no longer exists; mapper behavior is now exercised through the external-package tests in `resource_unmanaged_solution_test.go` (`package solution_test`) via the resource surface.

File: internal/services/solution/resource_unmanaged_solution.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+		return
+	}
+
+	dvExists, err := r.SolutionClient.DataverseExists(ctx, plan.EnvironmentId.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(fmt.Sprintf("Client error when checking if Dataverse exists in environment '%s'", plan.EnvironmentId.ValueString()), err.Error())
+		return
+	}
+	if !dvExists {
+		resp.Diagnostics.AddError(fmt.Sprintf("No Dataverse exists in environment '%s'", plan.EnvironmentId.ValueString()), "")
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/resource_unmanaged_solution.go]: Addressed - the diagnostic now carries a full detail explaining that unmanaged solutions require a Dataverse-enabled environment, that this environment has none provisioned, and what to verify before retrying.

File: internal/services/solution/datasource_unmanaged_solution.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	dvExists, err := d.SolutionClient.DataverseExists(ctx, state.EnvironmentId.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(fmt.Sprintf("Client error when checking if Dataverse exists in environment '%s'", state.EnvironmentId.ValueString()), err.Error())
+		return
+	}
+
+	if !dvExists {
+		resp.Diagnostics.AddError(fmt.Sprintf("No Dataverse exists in environment '%s'", state.EnvironmentId.ValueString()), "")
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/datasource_unmanaged_solution.go]: Addressed - the data source diagnostic now includes the equivalent detail (Dataverse not enabled/provisioned, verify the environment id, provision Dataverse first).

File: internal/services/solution/resource_unmanaged_solution.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+func validateUnmanagedSolution(solution *SolutionDto, typeName string) error {
+	if solution != nil && solution.IsManaged {
+		return fmt.Errorf("solution '%s' is managed and cannot be used with %s", solution.Name, typeName)
+	}
+
+	return nil
+}
+
+func splitSolutionCompositeID(id string) []string {
+	return strings.Split(id, "_")
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/resource_unmanaged_solution.go]: Addressed - `splitSolutionCompositeID` now uses `strings.SplitN(id, "/", 2)` (separator switched to `/` per the later review feedback, commit 86af4c94), so IDs are parsed into exactly two segments.

File: .changes/unreleased/added-20260330-165500-unmanaged-solution.yaml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+kind: added
+body: Added the `powerplatform_unmanaged_solution` resource for direct Dataverse solution CRUD without ZIP import lifecycle coupling.
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply .changes/unreleased/added-20260330-165500-unmanaged-solution.yaml:2]: I want to hold this PR's naming as `powerplatform_unmanaged_solution` - full reasoning in #1136 (my Apr 8 reply) and the summary comment on #1117. Short version: `powerplatform_solution` + `_import` names the mechanism rather than the state, and collapses two genuinely different state models. The managed counterpart is no longer hypothetical - #1218 implements `powerplatform_managed_solution` per #1143 - so the end-state pair is `unmanaged_solution`/`managed_solution` with the existing zip-import resource deprecating gracefully, no breaking rename needed. Happy to settle this in #1136.

File: internal/services/solution/api_solution.go [hunk 8]
```diff
@@ -258,6 +261,115 @@ func (client *Client) CreateSolution(ctx context.Context, environmentId string,
...
+		return nil, err
+	}
+	if err := client.Api.HandleNotFoundResponse(resp); err != nil {
+		return nil, err
+	}
+
+	return client.GetSolutionById(ctx, environmentId, solutionId)
+}
+
+func (client *Client) WaitForUnmanagedSolution(ctx context.Context, environmentId, solutionId, uniqueName string) (*SolutionDto, error) {
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/api_solution.go]: Done in 86af4c94 - unexported to `waitForUnmanagedSolution`.

File: internal/services/solution/api_solution.go [hunk 9]
```diff
@@ -258,6 +261,115 @@ func (client *Client) CreateSolution(ctx context.Context, environmentId string,
...
+	if createdSolution.Id != "" {
+		solutionID = createdSolution.Id
+	}
+	if entityId := resp.GetHeader("OData-EntityId"); entityId != "" {
+		if parsedSolutionID, ok := tryExtractSolutionIDFromEntityHeader(entityId); ok {
+			solutionID = parsedSolutionID
+		}
+	}
+
+	return client.WaitForUnmanagedSolution(ctx, environmentId, solutionID, uniqueName)
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/api_solution.go]: Done in 86af4c94 - the call site now uses the unexported `waitForUnmanagedSolution`.

File: internal/services/solution/datasource_unmanaged_solution.go (hunk 1 above)
Comment (AdamCoulterOz) [reply internal/services/solution/datasource_unmanaged_solution.go:3]: I would keep it: `powerplatform_solutions` returns the full list and callers must filter client-side, while this data source is a typed single-solution lookup by `environment_id` + `uniquename` that also validates the match is unmanaged and surfaces `id`/`publisher_id`/metadata directly for composition. That said, it is severable - if you feel strongly it duplicates, I can drop it from this PR and propose it separately.

File: internal/services/solution/resource_unmanaged_solution.go [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+	resp.TypeName = r.FullTypeName()
+	tflog.Debug(ctx, fmt.Sprintf("METADATA: %s", resp.TypeName))
+}
+
+func (r *UnmanagedSolutionResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	ctx, exitContext := helpers.EnterRequestContext(ctx, r.TypeInfo, req)
+	defer exitContext()
+
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "Resource for managing unmanaged Dataverse solutions as first-class solution records without coupling lifecycle to solution ZIP import operations.",
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/resource_unmanaged_solution.go:51]: Per the naming position (#1136), the description deliberately says *unmanaged* Dataverse solutions - the resource models the open, mutable solution container specifically, and the managed lifecycle now has its own resource in #1218. If the rename question resolves the other way we can revisit the wording together with the type name.

File: internal/services/solution/resource_unmanaged_solution.go [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+func (r *UnmanagedSolutionResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	ctx, exitContext := helpers.EnterRequestContext(ctx, r.TypeInfo, req)
+	defer exitContext()
+
+	var state UnmanagedSolutionResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/resource_unmanaged_solution.go:172]: Addressed - since merging current `main` (which includes #1173), the solution client maps an environment 404 to `ErrObjectNotFound`, and Read removes the resource from state when the environment is gone (Delete tolerates it as already-deleted). The Dataverse-not-enabled case surfaces an explicit diagnostic with detail rather than persisting state.

File: internal/services/solution/resource_unmanaged_solution.go [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+func (r *UnmanagedSolutionResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	ctx, exitContext := helpers.EnterRequestContext(ctx, r.TypeInfo, req)
+	defer exitContext()
+
+	idParts := splitSolutionCompositeID(req.ID)
+	if len(idParts) != 2 {
+		resp.Diagnostics.AddError(
+			"Invalid import ID",
+			fmt.Sprintf("Expected import ID in format 'environment_id_solution_id', got '%s'", req.ID),
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/resource_unmanaged_solution.go]: Done in 86af4c94 - the error text now reads `Expected import ID in format 'environment_id/solution_id'`.

File: internal/services/solution/resource_unmanaged_solution.go (hunk 2 above)
Comment (AdamCoulterOz) [reply internal/services/solution/resource_unmanaged_solution.go]: Done in 86af4c94 - `strings.SplitN(id, "/", 2)`.

File: examples/resources/powerplatform_unmanaged_solution/import.sh [hunk 13]
```diff
@@ -0,0 +1,2 @@
+# Unmanaged solution resources can be imported using the provider id format {environment_id}_{solution_id}
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply examples/resources/powerplatform_unmanaged_solution/import.sh]: Done in 86af4c94 - the comment now documents the `{environment_id}/{solution_id}` format.

File: examples/resources/powerplatform_unmanaged_solution/import.sh [hunk 14]
```diff
@@ -0,0 +1,2 @@
...
+terraform import powerplatform_unmanaged_solution.solution 00000000-0000-0000-0000-000000000000_00000000-0000-0000-0000-000000000001
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply examples/resources/powerplatform_unmanaged_solution/import.sh]: Done in 86af4c94 - the example now uses the `/` separator.

File: internal/services/solution/resource_unmanaged_solution_test.go [hunk 15]
```diff
@@ -0,0 +1,541 @@
...
+	"github.com/microsoft/terraform-provider-power-platform/internal/mocks"
+)
+
+const (
+	unmanagedSolutionEnvironmentID = "00000000-0000-0000-0000-000000000001"
+	unmanagedSolutionID            = "86928ed8-df37-4ce2-add5-47030a833bff"
+	unmanagedPublisherID           = "aa47dc6c-bf13-490b-a007-1da95a0d1e3f"
+)
+
+func TestUnitUnmanagedSolutionResource_Validate_Create(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/resource_unmanaged_solution_test.go:24]: The happy path has an acceptance version: `TestAccUnmanagedSolutionResource_Validate_Create_Update` covers create, metadata update, and the data source lookup against a real environment. I have not mirrored the error/boundary unit tests (invalid publisher, managed-solution mismatch, no-Dataverse) as acceptance tests - those need deliberately broken live environments, which does not fit the acceptance harness; the repo's definition of done also scopes acceptance tests to the happy path. Happy to add specific ones if you have a scenario in mind.

File: internal/services/solution/resource_unmanaged_solution_test.go [hunk 16]
```diff
@@ -0,0 +1,541 @@
...
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/jarcoal/httpmock"
+	"github.com/microsoft/terraform-provider-power-platform/internal/helpers"
+	"github.com/microsoft/terraform-provider-power-platform/internal/mocks"
+)
+
+const (
+	unmanagedSolutionEnvironmentID = "00000000-0000-0000-0000-000000000001"
+	unmanagedSolutionID            = "86928ed8-df37-4ce2-add5-47030a833bff"
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/resource_unmanaged_solution_test.go]: Done in 86af4c94 - fixtures and tests now use `00000000-0000-0000-0000-000000000002` (solution) and `...0003` (publisher).

File: internal/services/solution/resource_unmanaged_solution_test.go [hunk 17]
```diff
@@ -0,0 +1,541 @@
...
+		})
+
+	httpmock.RegisterResponder("POST", "https://00000000-0000-0000-0000-000000000001.crm4.dynamics.com/api/data/v9.0/solutions",
+		func(req *http.Request) (*http.Response, error) {
+			return httpmock.NewStringResponse(http.StatusNoContent, ""), nil
+		})
+
+	httpmock.RegisterResponder("GET", "https://00000000-0000-0000-0000-000000000001.crm4.dynamics.com/api/data/v9.2/solutions?%24expand=publisherid&%24filter=uniquename+eq+%27TerraformTestSolution%27",
+		func(req *http.Request) (*http.Response, error) {
+			readAttempts++
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/solution/resource_unmanaged_solution_test.go:213]: Addressed as far as the flow allows - all response payloads live in `.json` fixture files under `tests/resource/Validate_Unmanaged_*`. The remaining responder logic is a counter that serves `get_solution_before.json` then `get_solution_after.json` across the update sequence - the same sequenced-fixture pattern the repo's newer tests use (e.g. the #1173 `Read_ParentDeleted` tests), since a create-then-update flow genuinely returns different states over time.

File: internal/services/solution/datasource_unmanaged_solution.go (hunk 1 above)
Comment (AdamCoulterOz) [reply internal/services/solution/datasource_unmanaged_solution.go:3]: Following up after seeing the publishers data source direction on #1134 — happy to go with whatever you prefer here: drop the singular data source from this PR, or convert it to a plural `powerplatform_unmanaged_solutions` list form to match the convention. Say the word and I will adjust.

**Comment by AdamCoulterOz** [comment]
Parking this until the v5 train, as agreed with @mawasile: per the end state on #1136, this implementation returns as `powerplatform_solution` (the solution object) in v5, taking the name in the same major that removes the zip-import resource. Closing to keep the open-PR list matched to the current release; the branch stays live as the v5 source.
